cpp/src/text/detokenize.cu

```diff
@@ -78,61 +74,50 @@ struct detokenizer_fn {
       }
     }
-    return (nbytes > 0) ? (nbytes - d_separator.size_bytes()) : 0;
+    if (!d_chars) { d_offsets[idx] = (nbytes > 0) ? (nbytes - d_separator.size_bytes()) : 0; }
   }
 };
 
-/**
- * @brief Identifies indexes where the row value changes.
- */
-template <typename IndexType>
 struct index_changed_fn {
-  IndexType const* d_rows;
-  int32_t const* d_row_map;
-  __device__ bool operator()(cudf::size_type idx)
+  cudf::detail::input_indexalator const d_rows;
+  cudf::size_type const* d_row_map;
+  __device__ bool operator()(cudf::size_type idx) const
   {
     return (idx == 0) || (d_rows[d_row_map[idx]] != d_rows[d_row_map[idx - 1]]);
   }
 };
 
 /**
- * @brief This is a type-dispatch function to convert the row indices
- * into token offsets.
+ * @brief Convert the row indices into token offsets
+ *
+ * @param row_indices Indices where each token should land
+ * @param sorted_indices Map of row_indices sorted
+ * @param tokens_counts Token counts for each row
+ * @param stream CUDA stream used for kernel launches
  */
-struct token_row_offsets_fn {
-  cudf::column_view const row_indices;
-  cudf::column_view const sorted_indices;
-  cudf::size_type const tokens_counts;
-
-  template <typename T, std::enable_if_t<cudf::is_index_type<T>()>* = nullptr>
-  std::unique_ptr<rmm::device_uvector<cudf::size_type>> operator()(
-    rmm::cuda_stream_view stream) const
-  {
-    index_changed_fn<T> pfn{row_indices.data<T>(), sorted_indices.template data<int32_t>()};
-    auto const output_count =
-      thrust::count_if(rmm::exec_policy(stream),
-                       thrust::make_counting_iterator<cudf::size_type>(0),
-                       thrust::make_counting_iterator<cudf::size_type>(tokens_counts),
-                       pfn);
-    auto tokens_offsets =
-      std::make_unique<rmm::device_uvector<cudf::size_type>>(output_count + 1, stream);
-    thrust::copy_if(rmm::exec_policy(stream),
-                    thrust::make_counting_iterator<cudf::size_type>(0),
-                    thrust::make_counting_iterator<cudf::size_type>(tokens_counts),
-                    tokens_offsets->begin(),
-                    pfn);
-    // set the last element to the total number of tokens
-    tokens_offsets->set_element(output_count, tokens_counts, stream);
-    return tokens_offsets;
-  }
-
-  // non-integral types throw an exception
-  template <typename T, typename... Args, std::enable_if_t<not cudf::is_index_type<T>()>* = nullptr>
-  std::unique_ptr<rmm::device_uvector<cudf::size_type>> operator()(Args&&...) const
-  {
-    CUDF_FAIL("The detokenize indices parameter must be an integer type.");
-  }
-};
+rmm::device_uvector<cudf::size_type> create_token_row_offsets(
+  cudf::column_view const& row_indices,
+  cudf::column_view const& sorted_indices,
+  cudf::size_type tokens_counts,
+  rmm::cuda_stream_view stream)
+{
+  index_changed_fn pfn{cudf::detail::indexalator_factory::make_input_iterator(row_indices),
```

Comment:
I don't know what `pfn` means.  "P..." function?

Reply:
The 'p' is silent.